orttraining/orttraining/test/training_api/trainer/trainer.cc

```diff
@@ -0,0 +1,360 @@
+  ORT_RETURN_ON_ERROR(g_ort_training_api->CreateTrainingSession(env, soptions, checkpoint_state,
+                                                                params.model_training_graph_path.c_str(), do_eval ? params.model_evaluation_graph_path.value().c_str() : nullptr,
+                                                                params.optimizer_training_graph_path.size() > 0 ? params.optimizer_training_graph_path.c_str() : nullptr,
+                                                                &session));
+
+  size_t train_mode_output_count, eval_mode_output_count = 0;
+  ORT_RETURN_ON_ERROR(g_ort_training_api->TrainingSessionGetTrainModeOutputCount(session, &train_mode_output_count));
+
+  if (do_eval) {
+    ORT_RETURN_ON_ERROR(g_ort_training_api->TrainingSessionGetEvalModeOutputCount(session, &eval_mode_output_count));
+  }
+
+  int64_t sample_batch_count_per_epoch = 4;
+  if (sample_batch_count_per_epoch < params.train_batch_size || sample_batch_count_per_epoch % params.train_batch_size != 0) {
+    throw std::runtime_error("sample_count cannot be divisible by batch_size");
+  }
+  int64_t num_of_batches_per_epoch = sample_batch_count_per_epoch / params.train_batch_size;
+
+  onnxruntime::training::test::training_api::SyntheticDataLoader data_loader;
+  InitSyntheticDataLoader(data_loader, params, num_of_batches_per_epoch);
+
+  // TODO(baiju): Add C API for LRScheduler
+  // int64_t total_step_count = params.num_train_epochs * num_of_batches_per_epoch;
+  // int64_t warmup_step_count = total_step_count / 3;
+  // Ort::OrtLinearLRScheduler scheduler = Ort::OrtLinearLRScheduler(optimizer, warmup_step_count, total_step_count);
+
+  std::cout << "Initialization completed. Now starting training loop." << std::endl;
+  const int64_t stabilized_perf_start_step = 0;
+  double stabilized_total_end_to_end_time{0};
+  auto end_to_end_start = std::chrono::high_resolution_clock::now();
+
+  for (int64_t epoch = 0, batch_idx = 0; epoch < params.num_train_epochs; ++epoch) {
+    for (size_t step_in_cur_epoch = 0; step_in_cur_epoch < data_loader.NumOfSampleBatches(); ++step_in_cur_epoch) {
+      if (batch_idx >= stabilized_perf_start_step) {
+        end_to_end_start = std::chrono::high_resolution_clock::now();
+      }
+
+      std::vector<OrtValue*> inputs;
+      data_loader.GetNextSampleBatch(inputs);
+
+#if defined(USE_CUDA) && defined(ENABLE_NVTX_PROFILE)
+      onnxruntime::profile::NvtxRangeCreator train_step_range(
+          "module_TrainStep",
+          onnxruntime::profile::Color::Green);
+      train_step_range.Begin();
+#endif
+
+      std::vector<OrtValue*> fetches(train_mode_output_count);
+      ORT_RETURN_ON_ERROR(g_ort_training_api->TrainStep(session, nullptr,
+                                                        inputs.size(), (const OrtValue* const*)inputs.data(),
+                                                        train_mode_output_count, fetches.data()));
```

Comment:
is the cast necessary? if so, we should use C++ style casts
https://google.github.io/styleguide/cppguide.html#Casting